subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/BeanNode.java

```diff
@@ -19,32 +19,132 @@
 import com.google.common.base.Function;
 import com.google.common.base.Preconditions;
 import com.google.common.collect.Iterators;
+import org.gradle.api.Named;
 
 import javax.annotation.Nullable;
+import java.util.HashSet;
 import java.util.Iterator;
+import java.util.Map;
+import java.util.Set;
 
-public class BeanNode extends AbstractPropertyNode<BeanNode> {
-    private final Object bean;
+public abstract class BeanNode extends AbstractPropertyNode<BeanNode> {
 
-    public BeanNode(@Nullable String propertyName, Object bean) {
+    public static BeanNode create(@Nullable String propertyName, Object bean) {
+        if (bean instanceof Map<?, ?>) {
+            return new MapBeanNode(propertyName, (Map<?, ?>) bean);
+        }
+        if (bean instanceof Iterable<?>) {
+            return new IterableBeanNode(propertyName, (Iterable<?>) bean);
+        }
+        return new SimpleBeanNode(propertyName, bean);
+    }
+
+    protected BeanNode(@Nullable String propertyName, Object bean) {
         super(propertyName, Preconditions.checkNotNull(bean, "Null is not allowed as nested property '" + propertyName + "'").getClass());
-        this.bean = bean;
     }
 
-    public Object getBean() {
-        return bean;
+    public abstract Object getBean();
+
+    protected BeanNode createChildNode(String propertyName, @Nullable Object input) {
+        String qualifiedPropertyName = getQualifiedPropertyName(propertyName);
+        Object bean = Preconditions.checkNotNull(input, "Null is not allowed as nested property '" + qualifiedPropertyName + "'");
+        return BeanNode.create(qualifiedPropertyName, bean);
+    }
+
+    private static class SimpleBeanNode extends BeanNode {
+
+        private final Object bean;
+
+        public SimpleBeanNode(@Nullable String propertyName, Object bean) {
+            super(propertyName, bean);
+            this.bean = bean;
+        }
+
+        @Override
+        public Object getBean() {
+            return bean;
+        }
```

Comment:
This can be pulled up to `BeanNode`. You can even parametrize `BeanNode` to have typed access to this value in the derived types.

Reply:
I pulled it up to a `BaseClass`. I didn't want to pull it up to `BeanNode`, since clients of `BeanNode` do not care about the type parameter. If I would pull the type parameter to `BeanNode`, then it would need to be references as `BeanNode<?>` nearly everywhere.

Reply:
:+1: